tower-balance/src/load/peak_ewma.rs

```diff
@@ -0,0 +1,394 @@
+                rtt / NANOS_PER_MILLI,
+                self.rtt_ns / NANOS_PER_MILLI,
+            );
+            rtt
+        } else {
+            // When an RTT is observed that is less than the estimated RTT, we decay the
+            // prior estimate according to how much time has elapsed since the last
+            // update. The inverse of the decay is used to scale the estimate towards the
+            // observed RTT value.
+            let elapsed = nanos(now - self.update_at);
+            let decay = (-elapsed / decay_ns).exp();
+            let recency = 1.0 - decay;
+            let next_estimate = (self.rtt_ns * decay) + (rtt * recency);
+            trace!(
+                "update rtt={:03.0}ms decay={:06.0}ns; next={:03.0}ms",
+                rtt / NANOS_PER_MILLI,
+                self.rtt_ns - next_estimate,
+                next_estimate / NANOS_PER_MILLI,
+            );
+            next_estimate
+        };
+        self.update_at = now;
+
+        self.rtt_ns
+    }
+}
+
+// ===== impl Instrument =====
+
+impl Drop for Instrument {
+    fn drop(&mut self) {
+        let recv_at = clock::now();
+
+        if let Ok(mut rtt) = self.rtt_estimate.lock() {
+            if let Some(ref mut rtt) = *rtt {
+                rtt.update(self.sent_at, recv_at, self.decay_ns);
+                return;
+            }
+
+            *rtt = Some(RttEstimate::new(self.sent_at, recv_at));
+        }
+    }
+}
+
+// Utility that converts durations to nanos in f64.
+//
+// Due to a lossy transformation, the maximum value that can be represented is ~585 years,
+// which, I hope, is more than enough to represent request latencies.
+fn nanos(d: Duration) -> f64 {
+    const NANOS_PER_SEC: u64 = 1_000_000_000;
+    let n = f64::from(d.subsec_nanos());
+    let s = d.as_secs().saturating_mul(NANOS_PER_SEC) as f64;
+    n + s
+}
+
+#[cfg(test)]
+mod tests {
+    extern crate tokio_executor;
+    extern crate tokio_timer;
+
+    use futures::{Future, Poll, future};
+    use std::sync::{Arc, Mutex};
+    use std::time::{Duration, Instant};
+    use self::tokio_executor::enter;
+    use self::tokio_timer::clock;
+
+    use super::*;
+
+    struct Svc;
+    impl Service for Svc {
+        type Request = ();
+        type Response = ();
+        type Error = ();
+        type Future = future::FutureResult<(), ()>;
+
+        fn poll_ready(&mut self) -> Poll<(), ()> {
+            Ok(().into())
+        }
+
+        fn call(&mut self, (): ()) -> Self::Future {
+            future::ok(())
+        }
+    }
+
+    struct Now(Arc<Mutex<Instant>>);
+    impl clock::Now for Now {
+        fn now(&self) -> Instant {
+            *self.0.lock().expect("now")
+        }
+    }
+
+    #[test]
+    fn default_decay() {
+        let time = Arc::new(Mutex::new(Instant::now()));
+        let clock = clock::Clock::new_with_now(Now(time.clone()));
+
+        let mut enter = enter().expect("enter");
+        clock::with_default(&clock, &mut enter, |_| {
```

Comment:
@carllerche this works fine! though, ergonomically, this feels like boilerplate.  I think we could probably package up a default mock timer that handles this.